File: android/core/core-common/src/main/kotlin/app/kehdo/core/common/Outcome.kt
package app.kehdo.core.common

/**
 * Outcome of an operation that can succeed or fail.
 * Used everywhere instead of exceptions for predictable control flow.
 */
sealed interface Outcome<out T> {
    data class Success<out T>(val value: T) : Outcome<T>
    data class Failure(val error: KehdoError) : Outcome<Nothing>

    companion object {
        fun <T> success(value: T): Outcome<T> = Success(value)
        fun failure(error: KehdoError): Outcome<Nothing> = Failure(error)
    }
}

inline fun <T, R> Outcome<T>.map(transform: (T) -> R): Outcome<R> = when (this) {
    is Outcome.Success -> Outcome.Success(transform(value))
    is Outcome.Failure -> this
}

inline fun <T> Outcome<T>.onSuccess(block: (T) -> Unit): Outcome<T> {
    if (this is Outcome.Success) block(value)
    return this
}

inline fun <T> Outcome<T>.onFailure(block: (KehdoError) -> Unit): Outcome<T> {
    if (this is Outcome.Failure) block(error)
    return this
}

File: ios/Packages/KHCore/Sources/KHCore/Outcome.swift
import Foundation

/// Outcome of an operation that can succeed or fail.
/// Mirrors the Android `Outcome<T>` sealed class for cross-platform parity.
public enum Outcome<T> {
    case success(T)
    case failure(KehdoError)

    public func map<R>(_ transform: (T) -> R) -> Outcome<R> {
        switch self {
        case .success(let value): return .success(transform(value))
        case .failure(let error): return .failure(error)
        }
    }

    @discardableResult
    public func onSuccess(_ block: (T) -> Void) -> Outcome<T> {
        if case .success(let value) = self { block(value) }
        return self
    }

    @discardableResult
    public func onFailure(_ block: (KehdoError) -> Void) -> Outcome<T> {
        if case .failure(let error) = self { block(error) }
        return self
    }
}
